File: watch_tool.py
from pathlib import Path
import time
import os
from check_file_permission import check_file_permission

def watch_file(file_path, command, interval, config):
    if not check_file_permission(file_path, "read", config):
        raise PermissionError(f"Read access denied for {file_path}")
    file_path = Path(file_path).resolve()
    if not file_path.is_file():
        raise ValueError(f"File not found: {file_path}")
    last_mtime = file_path.stat().st_mtime
    while True:
        mtime = file_path.stat().st_mtime
        if mtime != last_mtime:
            os.system(command)
            last_mtime = mtime
        time.sleep(interval)

File: packages/core/src/tools/watch-tool.ts
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { promises as fs } from 'fs';
import { resolve } from 'path';
import { exec } from 'child_process';
import { checkFilePermission } from '../services/filePermissionService';

export async function watchFile(
  filePath: string,
  command: string,
  interval: number,
  config: unknown,
): Promise<void> {
  if (!(await checkFilePermission(filePath, 'read', config))) {
    throw new Error(`Read access denied for ${filePath}`);
  }

  const resolvedPath = resolve(filePath);
  const stats = await fs.stat(resolvedPath);

  if (!stats.isFile()) {
    throw new Error(`File not found: ${resolvedPath}`);
  }

  let lastMtime = stats.mtimeMs;

  const intervalId = setInterval(async () => {
    try {
      const currentStats = await fs.stat(resolvedPath);
      if (currentStats.mtimeMs !== lastMtime) {
        exec(command, (error, stdout, stderr) => {
          if (error) {
            console.error(`exec error: ${error}`);
            return;
          }
          console.log(`stdout: ${stdout}`);
          console.error(`stderr: ${stderr}`);
        });
        lastMtime = currentStats.mtimeMs;
      }
    } catch (error) {
      console.error(`Error watching file: ${error}`);
      clearInterval(intervalId);
    }
  }, interval * 1000);
}
